### ISSUE TITLE
Add logic operators on reputation criteria

### ISSUE THREAD

**Issue by cedoor** [bandada-infra/bandada#223]
## Description

Right now Bandada supports credential groups but a credential group supports only one [validator](https://github.com/bandada-infra/bandada/tree/main/libs/credentials/src/validators) (credential). Then you could create a group of people that have greater than or equal to x number of followers but can't check any other credential at the same time. It would be nice if Bandada can support groups with multiple credentials and logical operators. Then people will be able to create Bandada credential groups with many credentials like for example: `a group of people with greater than or equal to X number of Github followers (A) and contributed to a specific github repo (B) or have greater or equal to Y number of transactions (C)`, that would be the logical expression: `A and B or C`. It should also support parenthesis like for example: `A and (B or C)`. It should support any complex logical expression.

Logical operators to support:

- `AND`
- `OR`
- `NOT`
- `XOR`

Since it will support all of the above, it will consequently support `NAND`, `NOR` and `XNOR`.

It will support parentheses.

### Logical Operators

1. **AND**
   - **Symbol**: `&&` (in languages like C, C++, Java) or `AND` (in SQL)
   - **Description**: Returns true if both operands are true.
   - **Example**: `true && false` returns `false`.

2. **OR**
   - **Symbol**: `||` (in languages like C, C++, Java) or `OR` (in SQL)
   - **Description**: Returns true if at least one of the operands is true.
   - **Example**: `true || false` returns `true`.

3. **NOT**
   - **Symbol**: `!` (in languages like C, C++, Java) or `NOT` (in SQL)
   - **Description**: Returns true if the operand is false, and false if the operand is true.
   - **Example**: `!true` returns `false`.

4. **NAND** (Not AND)
   - **Symbol**: Often represented by `!(A && B)` since many languages don't have a direct NAND operator.
   - **Description**: Returns true if at least one of the operands is false.
   - **Example**: `!(true && true)` returns `false`.

5. **NOR** (Not OR)
   - **Symbol**: Often represented by `!(A || B)` since many languages don't have a direct NOR operator.
   - **Description**: Returns true if both operands are false.
   - **Example**: `!(true || false)` returns `false`.

6. **XOR** (Exclusive OR)
   - **Symbol**: `^` (in languages like C, C++, Java) or `XOR` (in SQL)
   - **Description**: Returns true if one, and only one, of the operands is true.
   - **Example**: `true ^ false` returns `true`.

7. **XNOR** (Exclusive NOR or Equivalence)
   - **Symbol**: Often represented by `!(A ^ B)` since many languages don't have a direct XNOR operator.
   - **Description**: Returns true if both operands are the same (both true or both false).
   - **Example**: `!(true ^ true)` returns `true`.

**Comment by 0xjei** [comment]
The `lib/credentials` package in Bandada defines providers such as Github, Blockchain, and EAS, each with one or more validators (e.g., GithubFollowers, GithubPersonalStars, BlockchainTransactions, and so on). Currently, each credential group only allows for the selection of one criteria (defined in the validator, e.g., "minNumberOfTransactions") that must be met to become a member and access the group. This is a huge limit, narrowing the set of supported use cases. 

To extend these, we need to work at the `validateCredentials()` method level and introduce a taxonomy of logical operators (e.g., `enum Logical Operators { AND: 'and', OR: 'or', ... }`).

My high level proposal is to add a loop in the current `validateCredentials()` method logic along with the combination of the fulfilment of `criteria[i-1]` with `criteria[i]`, based on the choosen logical operator. By creating a list of validators (criterias) and an ordered list of logical operators to combine them, it will be possible to handle complex use cases such as 'criterion1 && criterion2 || criterion3 && !criterion4'.

Creating a credential group with multiple criterias can be challenging from a UI/UX perspective, since it easily may become more complex and less intuitive.

FYI @vplasencia @aguzmant103

**Comment by aguzmant103** [comment]
> The `lib/credentials` package in Bandada defines providers such as Github, Blockchain, and EAS, each with one or more validators (e.g., GithubFollowers, GithubPersonalStars, BlockchainTransactions, and so on). Currently, each credential group only allows for the selection of one criteria (defined in the validator, e.g., "minNumberOfTransactions") that must be met to become a member and access the group. This is a huge limit, narrowing the set of supported use cases.
> 
> To extend these, we need to work at the `validateCredentials()` method level and introduce a taxonomy of logical operators (e.g., `enum Logical Operators { AND: 'and', OR: 'or', ... }`).
> 
> My high level proposal is to add a loop in the current `validateCredentials()` method logic along with the combination of the fulfilment of `criteria[i-1]` with `criteria[i]`, based on the choosen logical operator. By creating a list of validators (criterias) and an ordered list of logical operators to combine them, it will be possible to handle complex use cases such as 'criterion1 && criterion2 || criterion3 && !criterion4'.
> 
> Creating a credential group with multiple criterias can be challenging from a UI/UX perspective, since it easily may become more complex and less intuitive.
> 
> FYI @vplasencia @aguzmant103

Yes! Would this allow "( )" operators too? @0xjei

**Comment by 0xjei** [comment]
> > The `lib/credentials` package in Bandada defines providers such as Github, Blockchain, and EAS, each with one or more validators (e.g., GithubFollowers, GithubPersonalStars, BlockchainTransactions, and so on). Currently, each credential group only allows for the selection of one criteria (defined in the validator, e.g., "minNumberOfTransactions") that must be met to become a member and access the group. This is a huge limit, narrowing the set of supported use cases.
> > To extend these, we need to work at the `validateCredentials()` method level and introduce a taxonomy of logical operators (e.g., `enum Logical Operators { AND: 'and', OR: 'or', ... }`).
> > My high level proposal is to add a loop in the current `validateCredentials()` method logic along with the combination of the fulfilment of `criteria[i-1]` with `criteria[i]`, based on the choosen logical operator. By creating a list of validators (criterias) and an ordered list of logical operators to combine them, it will be possible to handle complex use cases such as 'criterion1 && criterion2 || criterion3 && !criterion4'.
> > Creating a credential group with multiple criterias can be challenging from a UI/UX perspective, since it easily may become more complex and less intuitive.
> > FYI [@vplasencia](https://github.com/vplasencia) [@aguzmant103](https://github.com/aguzmant103)
> 
> Yes! Would this allow "( )" operators too? [@0xjei](https://github.com/0xjei)

Yes, we can easily add a parser for derive the precedences given by the parentheses for the "equations".

**Comment by sripwoud** [comment]
question post merge due to me reviewing https://github.com/privacy-scaling-explorations/zk-kit/pull/310

Parsing logical operators sounds like a common problem, I am just curious why you haven't considered existing libraries/packages and chose to implement something from scratch? 
- bundle size?
- missing features?
- wanted something dependencies free?
- perf?

**Comment by vplasencia** [comment]
Hey @sripwoud! 

Most solutions I found were very old and/or used a tree data structure with parser and I decided that it's simpler to use just an algorithm that is mainly used with numbers and comparators but can be adapted to boolean expressions. It also solves parentheses well.

This algorithm parses and evaluates the expression at the same time which is better than parsing the expression and then using a tree, to evaluate it.

I didn't find any simpler and better implementation for our use case. 

Then in the future, the library could be extended to support numbers and comparators too. 

This algorithm will probably be implemented in contracts in the future if necessary for the Bandada on-chain credentials. This is another thing we had in mind, a tree like data structure wouldn't be a good option.
